app/commands/training_data/code_tags_sample/create_file_for_model_fine_tuning.rb

```diff
@@ -0,0 +1,413 @@
+      )
+    end
+
+    CRITERIA.each do |tag, criteria|
+      write_message!(
+        "When should a solution be tagged with the `#{tag}` EXERCISM_REPRESENTATION_TAG?",
+        "It should be assigned when the code #{criteria}."
+      )
+
+      write_message!(
+        "Which EXERCISM_REPRESENTATION_TAG should be tagged when the code #{criteria}?",
+        "`#{tag}`"
+      )
+    end
+  end
+
+  def write_messages_for_samples!
+    samples.each do |sample|
+      next if sample.code.blank?
+      next if sample.tags.blank?
+
+      write_messages_for_sample_tags!(sample)
+      write_messages_for_incorrect_llm_tags!(sample) if sample.llm_tags.present?
+      write_messages_for_missing_llm_tags!(sample) if sample.llm_tags.present?
+    rescue StandardError => e
+      Bugsnag.notify(e)
+    end
+  end
+
+  def samples
+    TrainingData::CodeTagsSample.includes(:track).
+      where(status: %i[community_checked admin_checked]).
+      where.not(tags: nil)
+  end
+
+  def write_messages_for_sample_tags!(sample)
+    instruction_text = <<~INSTRUCTION
+      Respond with a JSON object containing one top-level key called tags containing an array of EXERCISM_REPRESENTATION_TAGS (programming concepts, paradigms and techniques) for this #{sample.track.title} code.
+
+      ---
+
+      #{sample.code}"
+    INSTRUCTION
+
+    write_message!(
+      instruction_text.delete("\r"),
+      sample.tags.to_json
+    )
+  end
+
+  def write_messages_for_incorrect_llm_tags!(sample)
+    incorrect_llm_tags = sample.llm_tags - sample.tags
+    incorrect_llm_tags.each do |tag|
+      criteria = CRITERIA[tag]
+      next if criteria.blank?
+
+      criteria = criteria.split
+      criteria[0] = criteria[0].singularize
+      criteria = criteria.join(' ')
+
+      write_message!(
+        "Given the following code, should it be assigned the `#{tag}` EXERCISM_REPRESENTATION_TAG, and why?\n#{sample.code}",
+        "No, because it does not #{criteria}."
+      )
+    end
+  end
+
+  def write_messages_for_missing_llm_tags!(sample)
+    missing_llm_tags = sample.tags - sample.llm_tags
+    missing_llm_tags.each do |tag|
+      criteria = CRITERIA[tag]
+      next if criteria.blank?
+
+      write_message!(
+        "Given the following code, should it be assigned the `#{tag}` EXERCISM_REPRESENTATION_TAG, and why?\n#{sample.code}",
+        "Yes, because it #{criteria}."
+      )
+    end
+  end
+
+  def write_message!(user, assistant)
+    message = {
+      messages: [
+        { role: "system", "content": SYSTEM_TEXT },
+        { role: "user", "content": user },
+        { role: "assistant", "content": assistant }
+      ]
+    }
+
+    file.write(message.to_json)
+    file.write("\n")
+  end
+
+  def close_file!
+    file.close
+  end
+
+  def tags_with_namespace(namespace)
+    CRITERIA.keys.select { |tag| tag.starts_with?("#{namespace}:") }
+  end
+
+  SYSTEM_TEXT = 'You are a expert in EXERCISM_REPRESENTATION_TAGS'.freeze
+
+  CRITERIA = {
```

Comment:
Not sure if there is a better place for this 🤷

Reply:
I think these should all be in the database, probably in a global tag table.